src/main/kotlin/pet/project/app/service/impl/BookServiceImpl.kt

```diff
@@ -1,49 +1,64 @@
 import pet.project.app.annotation.Profiling
+import pet.project.app.dto.book.UpdateAmountRequest
 import pet.project.app.exception.BookNotFoundException
 import pet.project.app.model.Book
 import pet.project.app.repository.BookRepository
 import pet.project.app.service.BookService
-import kotlin.math.abs
+import pet.project.app.service.NotificationService
 
 @Profiling
 @Service
-class BookServiceImpl(private val bookRepository: BookRepository) : BookService {
+class BookServiceImpl(
+    private val bookRepository: BookRepository,
+    private val notificationService: NotificationService
+) : BookService {
 
-    override fun create(book: Book): Book = bookRepository.save(book)
+    override fun create(book: Book): Book = bookRepository.insert(book)
 
     override fun getById(bookId: String): Book =
         bookRepository.findByIdOrNull(bookId) ?: throw BookNotFoundException(bookId, "GET request")
 
     override fun update(book: Book): Book {
-        if (bookRepository.existsById(book.id!!.toHexString())) {
-            return bookRepository.save(book)
-        } else {
-            throw BookNotFoundException(book.id.toHexString(), "UPDATE request")
+        val updatedDocumentsCount = bookRepository.update(book)
+        if (updatedDocumentsCount != 1L) {
+            throw BookNotFoundException(book.id!!.toHexString(), "UPDATE request")
         }
+        return book
     }
 
-    override fun changeAmount(bookId: String, delta: Int): Int {
-        val book = getById(bookId)
-        val updatedAmount = (book.amountAvailable ?: 0) + delta
-        check(updatedAmount >= 0) { "Can't withdraw ${abs(delta)} book(s), when amount is ${book.amountAvailable}" }
-        val updatedBook = book.copy(amountAvailable = updatedAmount)
-        bookRepository.save(updatedBook)
-        if (updatedAmount == delta) {
-            log.info("Message for subscribers of book (id={})", bookId)
+    override fun updateAmount(request: UpdateAmountRequest): Boolean {
+        val modifiedCount = bookRepository.updateAmount(request)
+        if (modifiedCount != 1L) {
+            throw IllegalArgumentException("Requested book absent or have less amount available that needed: $request")
         }
-        return updatedAmount
+        if (request.delta > 0) {
+            notificationService.notifySubscribedUsers(request.bookId)
+        }
+        return true
+    }
+
+    override fun exchangeBooks(requests: List<UpdateAmountRequest>): Boolean {
+        val matchedCount = bookRepository.updateAmountMany(requests)
+        if (matchedCount != requests.size) {
+            throw IllegalArgumentException("Requested books absent or have less amount available that needed: $requests")
+        }
+
+        val booksWithIncreasedAmount = requests.filter { it.delta > 0 }
+        if (booksWithIncreasedAmount.isNotEmpty()) {
+            notificationService.notifySubscribedUsers(booksWithIncreasedAmount.map { it.bookId })
+        }
+
+        return true
```

Comment:
You don`t need to check if list is not empty 

```suggestion
    override fun exchangeBooks(requests: List<UpdateAmountRequest>): Boolean {
        val matchedCount = bookRepository.updateAmountMany(requests)
        if (matchedCount != requests.size) {
            throw IllegalArgumentException("Requested books absent or have less amount available that needed: $requests")
        }

        requests.filter { it.delta > 0 }
            .map { it.bookId }
            .let { notificationService.notifySubscribedUsers(it) }

        return true
```

Reply:
In this part, we are passing a List<String> instead of a String. I’ll rewrite this code to improve clarity.